website/blog/2024-08-19-otel-context-propagation-in-ferretdb.md

```diff
@@ -0,0 +1,163 @@
+
+In today's world of distributed systems, achieving reliability depends on various factors, one of which is effective observability.
+OpenTelemetry (OTel) has emerged as a standard for distributed tracing, but passing tracing data to databases remains a significant challenge,
+particularly with document databases like MongoDB.
+At FerretDB, we're committed to addressing this challenge.
+
+<!--truncate-->
+
+## OpenTelemetry context propagation for databases
+
+Context propagation, a concept that enables the tracking of requests as they move through different services, is explained by OpenTelemetry
+[here](https://opentelemetry.io/docs/concepts/context-propagation/).
+This context typically includes request-related data,
+such as trace identifiers, which are passed across service boundaries,
+allowing you to link different parts of a distributed request together.
+The most common approach to implement context propagation is by using [trace context](https://www.w3.org/TR/trace-context/)
+HTTP headers to pass this information between services, but this method is not always feasible.
+
+Obviously, most databases don't support such HTTP headers or other native ways to pass tracing-related context.
+Approaches like [SQLCommenter](https://google.github.io/sqlcommenter/) have been developed to bridge this gap,
+enabling the connection between current trace data and database queries.
+In these cases, trace information is injected at the ORM level and passed to the database via SQL comments.
+Database operators can then enable query logs to link specific queries with the corresponding application requests.
+
+But is it possible to achieve something similar with MongoDB's query language?
+At FerretDB, we believe that enabling the linking of a query's lifecycle with the caller's response can help developers make
+their applications more predictable and reliable.
+
+## FerretDB's approach to context propagation
+
+Let's consider a SQLCommenter-like approach for MongoDB.
+We can leverage MongoDB's existing features to attach metadata to queries.
+One such feature is the comment field, which allows additional information to be included in MongoDB queries.
+This field doesn't influence query logic but can be used to provide valuable context.
+
+At FerretDB, we want to empower users to pass such context.
+We decided to utilize the `comment` field to do it.
+We parse the content of the `comment` field, and if it's a JSON document with a `ferretDB` key, we check for tracing data.
+If the tracing data is present, FerretDB sets the parent context for a span created for the current operation.
+
+An example of a comment with tracing data could look like this:
+
+```json
+{
+  "ferretDB": {
+    "traceID": "1234567890abcdef1234567890abcdef",
+    "spanID": "fedcba9876543210"
+  }
+}
+```
+
+The strings `traceID` and `spanID` are hex-encoded strings that represent the corresponding trace and span identifiers.
```

Comment:
I think `hex-encoded strings` is no longer the case from above examples. Unless its about how it's implemented in FerretDB :thinking: